src/prefect/engine.py

```diff
@@ -1616,3 +1635,47 @@ def link_if_trackable(obj: Any) -> None:
         )
         # Let the exit code be determined by the base exception type
         raise
+
+
+@asynccontextmanager
+async def watch_for_flow_run_cancellation(
+    flow_run: FlowRun, task_group: anyio.abc.TaskGroup, client: "OrionClient", logger
+):
+    async def _worker(
+        flow_run: FlowRun,
+        flow_run_cancel_scope: anyio.CancelScope,
+        task_status: anyio.abc.TaskStatus,
+    ):
+        """
+        This worker looks at the flow run state on an interval. if the flow run is
+        CANCELLED it will cancel the scope that has been passed to it which should
+        interrupt any code within the `watch_for_flow_run_cancellation` context.
+
+        If the state is terminal, this worker will exit.
+        If the `watch_for_flow_run_cancellation` context exits, this worker should be
+        interrupted by the scope it returns on start.
+        """
+
+        with anyio.CancelScope() as worker_cancel_scope:
+            task_status.started(worker_cancel_scope)
+            while not flow_run.state.is_final():
+                await anyio.sleep(PREFECT_CANCEL_CHECK_QUERY_INTERVAL.value())
+                flow_run = await client.read_flow_run(flow_run.id)
+                if flow_run.state.type == StateType.CANCELLED:
+                    logger.info(
+                        "Detected cancellation of flow run! Sending cancellation..."
+                    )
+                    flow_run_cancel_scope.cancel()
+
+    # Create a scope that will allow cancellation of the code that occurs during yield
+    with anyio.CancelScope() as flow_run_cancel_scope:
```

Comment:
We'd probably want to refactor to declare this scope on the flow run context instead which would be useful for CRASH monitoring to know that it was an intentional interrupt.